enable njit for count method on unicode type

jdburnet opened the pull request:
Use the find method overload as a reference which uses naive string
matching. Add test cases that closely follow those of the find method as
well. However, add scenarios when testing the count method on literals
to fully cover comparison code paths.

jdburnet committed:
enable njit for count method on unicode type

Use the find method overload as a reference which uses naive string
matching. Add test cases that closely follow those of the find method as
well. However, add scenarios when testing the count method on literals
to fully cover comparison code paths.

jdburnet commented:
travis-ci build is timing out for some jobs, see numba/numba#3391

jdburnet commented:
double checked that all tests in the test class that this PR modifies are passing,
```bash
(numbaenv) root@JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python -m numba.runtests numba.tests.test_unicode.TestUnicode -v
test_basic_gt (numba.tests.test_unicode.TestUnicode) ... ok
test_basic_lt (numba.tests.test_unicode.TestUnicode) ... ok
test_comparison (numba.tests.test_unicode.TestUnicode) ... ok
test_concat (numba.tests.test_unicode.TestUnicode) ... ok
test_count (numba.tests.test_unicode.TestUnicode) ... ok
test_endswith (numba.tests.test_unicode.TestUnicode) ... ok
test_eq (numba.tests.test_unicode.TestUnicode) ... ok
test_find (numba.tests.test_unicode.TestUnicode) ... ok
test_ge (numba.tests.test_unicode.TestUnicode) ... ok
test_getitem (numba.tests.test_unicode.TestUnicode) ... ok
test_getitem_error (numba.tests.test_unicode.TestUnicode) ... ok
test_gt (numba.tests.test_unicode.TestUnicode) ... ok
test_in (numba.tests.test_unicode.TestUnicode) ... ok
test_inplace_concat (numba.tests.test_unicode.TestUnicode) ... ok
test_join (numba.tests.test_unicode.TestUnicode) ... ok
test_join_empty (numba.tests.test_unicode.TestUnicode) ... ok
test_join_interleave_str (numba.tests.test_unicode.TestUnicode) ... ok
test_join_non_string_exception (numba.tests.test_unicode.TestUnicode) ... ok
test_le (numba.tests.test_unicode.TestUnicode) ... ok
test_len (numba.tests.test_unicode.TestUnicode) ... ok
test_literal (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_comparison (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_concat (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_count (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_find (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_getitem (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_in (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_len (numba.tests.test_unicode.TestUnicode) ... ok
test_literal_xyzwith (numba.tests.test_unicode.TestUnicode) ... ok
test_lt (numba.tests.test_unicode.TestUnicode) ... ok
test_passthrough (numba.tests.test_unicode.TestUnicode) ... ok
test_pointless_slice (numba.tests.test_unicode.TestUnicode) ... ok
test_repeat (numba.tests.test_unicode.TestUnicode) ... ok
test_repeat_exception_float (numba.tests.test_unicode.TestUnicode) ... ok
test_slice2 (numba.tests.test_unicode.TestUnicode) ... ok
test_slice2_error (numba.tests.test_unicode.TestUnicode) ... ok
test_slice3 (numba.tests.test_unicode.TestUnicode) ... ok
test_slice3_error (numba.tests.test_unicode.TestUnicode) ... ok
test_split (numba.tests.test_unicode.TestUnicode) ... ok
test_split_exception_empty_sep (numba.tests.test_unicode.TestUnicode) ... ok
test_split_exception_noninteger_maxsplit (numba.tests.test_unicode.TestUnicode) ... ok
test_split_whitespace (numba.tests.test_unicode.TestUnicode) ... ok
test_split_with_maxsplit (numba.tests.test_unicode.TestUnicode) ... ok
test_startswith (numba.tests.test_unicode.TestUnicode) ... ok
test_stride_slice (numba.tests.test_unicode.TestUnicode) ... ok
test_walk_backwards (numba.tests.test_unicode.TestUnicode) ... ok
test_zfill (numba.tests.test_unicode.TestUnicode) ... ok

----------------------------------------------------------------------
Ran 47 tests in 16.623s

OK
```

Alexander-Makaryev commented:
Hello,
Thank you for solution.
Please, look into the example below.
```python
import numba
def python_fun(a):
    aa = a.count("22")
    return aa
numba_fun = numba.njit(python_fun)
s = "122222234"
print("s = ", s, ", python_fun(s) = ", python_fun(s), ", numba_fun(s) = ", numba_fun(s))

#s =  122222234 , python_fun(s) =  3 , numba_fun(s) =  5
```

jdburnet committed:
add failing test case for unicode count

Add failing unit test where unicode count method was double counting
patterns made of repeated characters.

jdburnet committed:
fix failing test case for unicode count

Fix bug when matching a pattern of repeated characters in the unicode
count method

jdburnet commented:
@Alexander-Makaryev, thanks for pointing out that bug. I've made two more commits, one to add a test casing exposing the same bug and another to fix the test

oleksandr-pavlyk commented on numba/unicode.py:
```diff
@@ -377,11 +377,18 @@ def _find(substr, s):
     count = 0
-    for i in range(len(s) - len(substr) + 1):
+    i = 0
+    incr = len(substr)
```
Does saving `len(substr)` in a variable improve performance ?

jdburnet replied:
@oleksandr-pavlyk, I'm understanding your question as, if we were to generally save `len(substr)` in a variable (e.g. `substrlen`) at the beginning of the function would we improve performance of the whole function? If that's not right please correct me.

In the context of a jit wrapped function the `len()` call is just an access to the `_len` attribute. So if we saved the length of the substring in a local variable then we could save us the overhead of a function call and a dictionary look up at the expense of using a little extra memory. Since the user is most likely using the jit decorator to gain maximum performance, a little extra memory is most likely a trade off we are willing to make.

If you agree, I will go ahead and make that change

oleksandr-pavlyk replied:
Yes, I had in mind saving the result in a variable, and reusing it. I think it's worth checking how much of a performance impact this change actually has, if only to gain an insight into how much numba is able to optimize on its own.

jdburnet replied:
Got it, agreed. I will put together some comparisons with `timeit` and report back

jdburnet replied:
I put together this experiment,
```python
import numba
import timeit


@numba.njit
def compiledcount(a, b):
    return a.count(b)


PATTERNS = (
    "aaaaaa",
    "aa"
    "abcdef",
    "cde",
)
search_string = ''.join(random.choice(PATTERNS) for _ in range(1000))
target = random.choice(PATTERNS)
NUM = 1000000


comptime = timeit.timeit(
    'compiledcount(search_string, target)',
    globals=globals(),
    number=NUM
)
print(comptime, "seconds to perform", NUM, "loops of compiled count")
print(comptime/NUM, "seconds per loop")
```
Here is the implementation of count that does not save off len(substr) into a variable
```python
@njit(_nrt=False)
def _count(substr, s):
    count = 0
    i = 0
    incr = len(substr)
    if not incr:
        incr = 1
    # Naive, slow string matching for now
    while i < len(s) - len(substr) + 1:
        if _cmp_region(s, i, substr, 0, len(substr)) == 0:
            count += 1
            i += incr
        else:
            i += 1
    return count
```
Using this implementation to run the script above,
```bash
(numbaenv) JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
35.33945925699663 seconds to perform 1000000 loops of compiled count
3.533945925699663e-05 seconds per loop
```
Here is the count implementation _with_ saving len(substr) into a local variable,
```python
@njit(_nrt=False)
def _count(substr, s):
    count = 0
    i = 0
    substrlen = len(substr)
    incr = substrlen
    if not incr:
        incr = 1
    # Naive, slow string matching for now
    while i < len(s) - substrlen + 1:
        if _cmp_region(s, i, substr, 0, substrlen) == 0:
            count += 1
            i += incr
        else:
            i += 1
    return count
```
Using this implementation,
```bash
(numbaenv) JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
37.131292551988736 seconds to perform 1000000 loops of compiled count
3.713129255198874e-05 seconds per loop
```
So, you can see that, for this particular experiment, not saving len(substr) into a local variable was actually slightly faster. At this point I'm not sure if this would hold for all workloads.

As a point of reference I tried running the same script but with the njit decorator removed,
```bash
(numbaenv) root@JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
3.477520063999691 seconds to perform 1000000 loops of python count
3.477520063999691e-06 seconds per loop
```
Interestingly, removing the njit decorator and running the pure python count implementation gives us an order of magnitude speed up. Either this experiment is not demonstrating a good numba use case or the count implementation just needs to be sped up. When I have a bit of time I will go into CPython and see what their string matching looks like.

cc @oleksandr-pavlyk

jdburnet replied:
[As expected, the string matching in CPython is far fro naive (click me)](https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Objects/stringlib/fastsearch.h)

I reran the experiment using a search string that is ~10 characters long rather than thousands of characters. Relative performance of CPython versus numba jit compiled is still an order of magnitude difference in favor of CPython.
```bash
(numbaenv) JB-rhel ~/personal/repos/numba (feature/unicode-type-count) $ python count_times.py
0.1575793359952513 seconds to perform 1000000 loops of python count
1.575793359952513e-07 seconds per loop
3.9455732989881653 seconds to perform 1000000 loops of compiled count
3.945573298988165e-06 seconds per loop
```
